db/versions/0017.yml

```diff
@@ -0,0 +1,166 @@
+        returning denylisted_notification.etag;
+      end
+  denylisted_notification_entities_scan:
+    description: See taskcluster-lib-entities
+    mode: read
+    serviceName: notify
+    args: pk text, rk text, condition text, size integer, page integer
+    returns: table (partition_key text, row_key text, value jsonb, version integer, etag uuid)
+    body: |-
+      declare
+        cond text[];
+        exp_cond_operator text;
+        exp_cond_operand timestamptz;
+        partition_key_var text;
+        row_key_var text;
+      begin
+        if not condition is null then
+          raise exception 'no condition support in denylisted_notification_entities_scan';
+        else
+          return query select
+            encode_string_key(denylisted_notification.notification_type) as partition_key,
+            encode_string_key(denylisted_notification.notification_address) as row_key,
+            jsonb_build_object(
+              'PartitionKey', encode_string_key(denylisted_notification.notification_type),
+              'RowKey', encode_string_key(denylisted_notification.notification_address),
+              'notificationType', notification_type,
+              'notificationAddress', notification_address
+            ) as value,
+            1 as version,
+            denylisted_notification.etag as etag from denylisted_notification
+          where
+            (denylisted_notification_entities_scan.pk is null or decode_string_key(denylisted_notification_entities_scan.pk) = notification_type) and
+            (denylisted_notification_entities_scan.rk is null or decode_string_key(denylisted_notification_entities_scan.rk) = notification_address)
+          order by denylisted_notification.notification_type, denylisted_notification.notification_address
+          limit case
+            when (size is not null and size > 0) then size + 1
+            else null
+          end
+          offset case
+            when (size is not null and size > 0 and page is not null and page > 0) then page
+            else 0
+          end;
+        end if;
+      end
+  add_denylist_address:
+    description: |-
+      If the denylist address already exists, this is a no-op. Otherwise, add the denylist
+      address for the taskcluster-notify service, with a new random etag.
+    mode: write
+    serviceName: notify
+    args: address_in jsonb
+    returns: void
+    body: |-
+      begin
+        insert into denylisted_notification(notification_type, notification_address, etag)
+        values (
+          (address_in ->> 'notificationType')::text,
+          (address_in ->> 'notificationAddress')::text,
+          public.gen_random_uuid()
+        ) on conflict do nothing;
+      end
+  delete_denylist_address:
+    description: |-
+      Delete a denylist address for the taskcluster-notify service.
+      Returns number of rows deleted (0 or 1).
+    mode: write
+    serviceName: notify
+    args: address_in jsonb
```

Comment:
This is sort of a shortcut to named arguments.  But when we add named arguments, it will need to be called as `db.fns.delete_denylist_address({address_in: {notificationType: 'abc', notificationAddress: 'def'}})`.  Is that desirable?

Reply:
I've changed it, but in general I'm not sure there is value in mapping all fields to unique database columns, unless those columns join to other columns in database queries. Having the data structures exploded means the database and the application need to keep track of the data structures, rather than them just being managed by the json schemas, so leads to additional boilerplate and introduces risk of divergence between the database model and the jsonschema definitions.

Reply:
OK, I've changed it.